pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,367 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	azureclient "github.com/gardener/gardener-extension-provider-azure/pkg/azure/client"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-03-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-05-01/network"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	ctrlerror "github.com/gardener/gardener/extensions/pkg/controller/error"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// bastionEndpoints collects the endpoints the bastion host provides; the
+// private endpoint is important for opening a port on the worker node
+// ingress NSG rule to allow SSH from that node, the public endpoint is where
+// the enduser connects to establish the SSH connection.
+type bastionEndpoints struct {
+	private *corev1.LoadBalancerIngress
+	public  *corev1.LoadBalancerIngress
+}
+
+// Ready returns true if both public and private interfaces each have either
+// an IP or a hostname or both.
+func (be *bastionEndpoints) Ready() bool {
+	return be != nil && IngressReady(be.private) && IngressReady(be.public)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	var factory = azureclient.NewAzureClientFactory(a.client)
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	publicIP, err := ensurePublicIPAddress(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	nic, err := ensureNic(ctx, factory, opt, publicIP)
+	if err != nil {
+		return err
+	}
+
+	opt.NicID = *nic.ID
```

Comment:
Currently the call stack looks like this:
`ensureComputeInstance`:`5 arguments` -> `computeInstanceDefine`:`3 agruments` -> `ID: &opt.NicID`
If we pass it as a separate argument it will be 6 and 4 arguments respectively. `6` looks too much from my point of view...
Tell me if you have a strong opinion about this then I will change.